### ISSUE TITLE
captureStreamUntilEnded(): consider clarifying or removing.

### ISSUE THREAD

**Issue by yell0wd0g** [w3c/mediacapture-fromelement#42]
Method [captureStreamUntilEnded()](https://w3c.github.io/mediacapture-fromelement/#widl-HTMLMediaElement-captureStreamUntilEnded-MediaStream) essentially states that when `<video>`/`<audio>` element finishes playback and signals it via firing an [`ended`](https://html.spec.whatwg.org/multipage/embedded-content.html#event-media-ended) event, the  captured `MediaStream` should stop all its `Tracks`. 

This is something that the WebApp can also do that via listening to the said `<video>`/`<audio>` `ended` event and subsequently calling [`stop()`](https://www.w3.org/TR/mediacapture-streams/#widl-MediaStreamTrack-stop-void) in each of the captured `MediaStream` `Tracks`. I propose removing it from the Spec, or clarifying what's its added value.

(As a personal note, perhaps this method predates the existence of [`MediaStreamTrack.stop()`](https://www.w3.org/TR/mediacapture-streams/#widl-MediaStreamTrack-stop-void) which would justify both its existence and its removal).

@foolip, since this question arose during a Blink code review.

Just FTR, the method is implemented in Gecko ([here](https://github.com/mozilla/gecko-dev/blob/42c9a619ec0da7e40b5d848f8354c06d5a4747ee/dom/html/HTMLMediaElement.cpp#L2274)).

**Comment by alvestrand** [comment]
My memory says that the spec was written based on the Firefox implementation. I think the purpose of the function was to provide "fire-and-forget" capture, where you didn't have to do anything explicit to stop the recording when the media element was stopped.

If you only have captureStream(), you have to add an event listener for the ended event on the media element, and stop the tracks from that event listener.

**Comment by foolip** [comment]
OK, there is a [`mozCaptureStreamUntilEnded()` in HTMLMediaElement.idl](https://github.com/mozilla/gecko-dev/blob/5b76709f3e84929482be34f47d7a0482224111ef/dom/webidl/HTMLMediaElement.webidl#L128), so there's actually an implementation, albeit prefixed.

When reviewing https://codereview.chromium.org/2110573002/ I wondered about the precise timing, getting interoperable implementations that all send all of the old resource but none of the new resource through the MediaStream seems tricky on the face of it, and probably isn't guaranteed by a simple implementation that follows the timing of the ended event.

Maybe one of the Gecko implementers can comment on how they envision this being used and whether they think removal would acceptable?

**Comment by martinthomson** [comment]
I think that part of the original thinking was to address the time shear between the media processing and the script threads.  You only receive 'ended' some (small) time after playback ceases and any instruction to terminate recording will take some time to propagate.  That means that you could have dead air on a stream.

/cc @Pehrsons who probably has more relevant info.

**Comment by Pehrsons** [comment]
This implementation predates me working on gecko, but I think the reasoning is more or less what @martinthomson wrote.

Calling `stop()` on tracks in the `ended` handler is not a proper polyfill, since current tracks have already stopped and you cannot prevent tracks from a new source from appearing in the stream.

**Comment by foolip** [comment]
In an attempt to understand what is intended, I took a look at Gecko's implementation, and I think these are the two important parts timing-wise:
- [`HTMLMediaElement::PlaybackEnded`](https://github.com/mozilla/gecko-dev/blob/fc8e03f0d8491ad0c6380e3832087bda05f75e90/dom/html/HTMLMediaElement.cpp#L4057)
- [`TrackUnionStream::ProcessInput`](https://github.com/mozilla/gecko-dev/blob/fc8e03f0d8491ad0c6380e3832087bda05f75e90/dom/media/TrackUnionStream.cpp#L136)

Notably, the bit in `HTMLMediaElement::PlaybackEnded` seems to no match [the spec](https://w3c.github.io/mediacapture-fromelement/#dfn-capturestreamuntilended), which says "The resulting stream ends when the media element has [ended playback](https://html.spec.whatwg.org/multipage/embedded-content.html#ended-playback)", where "ended playback" isn't true if there's a `loop` attribute while playing forwards, but Gecko does the same thing even if looping. Spec bug or implementation bug?

Anyway, I'm not sure if this really needed to end the stream before tracks from a new media resource appear. Once a media element is in `readyState` `HAVE_METADATA` or higher, the only thing that can cause a new resource to be loaded is scripts. So, if one wants to end the stream before the new resource begins, can't one simply end all the tracks before loading the new resource and/or disconnect the stream from whatever it was attached to?

**Comment by Pehrsons** [comment]
Note that gecko's implementation is old and not up to spec. This is why it is still prefixed. Seeking during capture will for instance end all tracks and create new ones.

If capture is done by an extension (WebExtension in Firefox, same model as in Chrome) it won't have control over when content loads a new resource.
Could content change the source (perhaps to an active MediaStream that doesn't rely on network I/O to load metadata) so that the media processing thread exposes new tracks before the extension has seen the "ended" event?

**Comment by foolip** [comment]
Are there Firefox extensions that use `captureStreamUntilEnded()` for the timing guarantees?

Let's see what timing guarantees the spec makes. Currently, if an extension script calls `captureStreamUntilEnded()` and the content script changes the `src` attribute mid-playback, then:
1. The `src` attribute setter invokes the [media element load algorithm](https://html.spec.whatwg.org/multipage/embedded-content.html#media-element-load-algorithm),
2. which queues an `abort` event and synchronously invokes the [resource selection algorithm](https://html.spec.whatwg.org/multipage/embedded-content.html#concept-media-load-algorithm), which:
   1. awaits a stable state (i.e. waits for a microtask checkpoint),
   2. then queues a task to fire a `loadstart` event,
   3. then "End the synchronous section, continuing the remaining steps in parallel",
   4. then invokes the [resource fetch algorithm](https://html.spec.whatwg.org/multipage/embedded-content.html#concept-media-load-resource),
3. which can "queue a task to run the first appropriate steps from the [media data processing steps list](https://html.spec.whatwg.org/multipage/embedded-content.html#media-data-processing-steps-list)"
4. which is where new audio tracks and video tracks are added.

In other words, the spec seems to guarantee that the `abort` and `loadstart` events will be fired (not just queued) before any new tracks have been added, which should allow the extension script to stop using the captured media stream. Mid-playback changes should be the hardest case, so I think that covers it.

It's quite likely that implementations don't actually match the exact timing of the spec here, but checking if that can be fixed seems best before adding APIs to provide timing guarantees.

**Comment by foolip** [comment]
Another possibility is that the timing of `AudioTrack`/`VideoTrack` objects being added to `audioTracks` and `videoTracks` lists (where implemented) in the "media data processing steps list" is different from the timing of `MediaStreamTrack` objects appearing in the `MediaStream` returned from `captureStream*()`, but that also seems like an unfortunate situation that would be worth trying to rectify.

**Comment by Pehrsons** [comment]
Right, thanks for the digging. Then removing `captureStreamUntilEnded` is fine by me.

**Comment by yell0wd0g** [comment]
@foolip, @martinthomson, following @Pehrsons comment I guess it's OK to remove the spec parts related to `captureStreamUntilEnded()`?

**Comment by foolip** [comment]
That is what I hoped for, yes, less API surface to understand, implement and write tests for.

**Comment by foolip** [comment]
Can this issue be closed now?
